### ISSUE TITLE
Add animation support

### ISSUE THREAD

**Issue by jason-bowen-zheng** [pyglet/pyglet#772]
The title should be clearly explained: the animation means that the program smoothly change **an object's attribute** from `A` to `B` in a specify time.
> Specifically it is very much like CSS animation.

This feature should be available on sprite, shape, gui widget and label by inherite a `Animatable` class.

We can also have an `AnimationGroup` to create complex animation.

**Comment by benmoran56** [comment]
(This is a follow up conversation from https://github.com/pyglet/pyglet/pull/466)

Now I understand what you mean. The example I posted is very low-level, and needs to be subclassed into more useful things. 

At the lowest level, I am thinking of Animations like this:
**Frame**:  Consists of "data", and "time".  Data could be position, image, 3D model vertices, etc.
**Animation**:  A collection of Frames. 
**AnimationController**: Moves the Animation between frames, in some specific way. 

In pyglet, we already have image animation:
**Frame**:  image, time until next image
**Animation**:  list of Frames
**AnimationController**: simply change the image when the time is reached. 

Other types of animation can be made from the same parts. There are many types of
movement, such as linear, lerp, etc. Also, 3D models can have animations with different lists of vertices.
The AnimationController will determine how the Frames are switched between. 

There are many possibilities, but some commone ones can be pre-defined so it's easier for users.
I made a simple example of a MovementAnimator using lerp (linear interpolation):

```py
import pyglet

from pyglet import clock
from pyglet.math import Vec2
from pyglet.animation import AnimationController, Animation, AnimationFrame

window = pyglet.window.Window(960, 540)
batch = pyglet.graphics.Batch()
circle = pyglet.shapes.Circle(100, 100, 25, color=(200, 50, 50), batch=batch)


frame1 = AnimationFrame(data=Vec2(100, 100), duration=3)    # position 1
frame2 = AnimationFrame(data=Vec2(500, 300), duration=6)    # position 2
frame3 = AnimationFrame(data=Vec2(300, 500), duration=7)    # position 3

animation = Animation(frames=[frame1, frame2, frame3])


class MovementAnimator(AnimationController):

    def __init__(self, shape, animation):
        self.shape = shape
        self.rate = 1/60

        self._animation = animation
        clock.schedule_interval(self._animate, self.rate)

    def _animate(self, dt):
        frame = self._animation.frames[self._frame_index]

        current = Vec2(*self.shape.position)
        target = frame.data

        if current.distance(target) < 0.1:
            self._frame_index += 1
        else:
            self.shape.position = current.lerp(target, dt * frame.duration)

        if self._frame_index >= len(self._animation.frames):
            self._frame_index = 0
            self.dispatch_event('on_animation_end')


shape_animator = MovementAnimator(shape=circle, animation=animation)


@window.event
def on_draw():
    window.clear()
    batch.draw()


pyglet.app.run()
```

**Comment by jason-bowen-zheng** [comment]
Using the above example to create animations seems complicated.

Suppose the `Circle` inherite a class which provide animation operation:
```python
circle = pyglet.shapes.Circle(100, 100, 25, color=(200, 50, 50))
animate = circle.animate("position", (500, 300), duration=6, model="lerp").animation("r", 50, duration=7)
animate.run()
```
which can move the circle from `(100, 100)` to `(500, 300)` and change radius.

However, the disadvantage is that we cannot control whether the animations are executed in parallel or one after the other.

**Comment by Kiera38** [comment]
It's possible to add an Animator class to pyglet.animation like this. This works similar to @benmoran56's example, but allows you to animate any attribute of the object:
```py
from pyglet.math import Vec2, Vec3, Vec4
from pyglet import clock


def linear(progress):
    return progress
    

def interpolate_number(start, target, progress):
    return start + (target - start) * progress
    

def interpolate_vec2(start, target, progress):
    return Vec2(*start).lerp(Vec2(*target), progress)
    

def interpolate_vec3(start, target, progress):
    return Vec3(*start).lerp(Vec3(*target), progress)
    

def interpolate_vec4(start, target, progress):
    return Vec4(*start).lerp(Vec4(*target), progress)


class Animator(AnimationController):
    def __init__(self, obj, attr, animation, rate=1/60, easing=linear):
        self._obj = obj
        self._attr = attr
        self._animation = animation
        self._rate = rate
        self._easing = easing
        
        self._start = getattr(obj, attr)
        if isinstance(self._start, (int, float)):
            self._interpolate = interpolate_number
        elif len(self._start) == 2:
            self._interpolate = interpolate_vec2
        elif len(self._start) == 3:
            self._interpolate = interpolate_vec3
        elif len(self._start) == 4:
            self._interpolate = interpolate_vec4
            
        self._frame_time = 0
        clock.schedule_interval(self._animate, rate)
        
    def _animate(self, dt):
        frame = self._animation.frames[self._frame_index]
        
        current = self._interpolate(self._start, frame.data, self._easing(self._frame_time / frame.duration))
        setattr(self._obj, self._attr, current)
        self._frame_time += dt
        
        if self._frame_time >= frame.duration:
            self._frame_index += 1
            self._frame_time = 0
            self._start = current
            
            if self._frame_index >= len(self._animation.frames):
                self._frame_index = 0
                self.dispatch_event('on_animation_end')
```
Here is an example:
```py
import pyglet

from pyglet.animation import Animator, Animation, AnimationFrame

window = pyglet.window.Window(960, 540)
batch = pyglet.graphics.Batch()
circle = pyglet.shapes.Circle(100, 100, 20, color=(200, 50, 50), batch=batch)

animation = Animation(frames=[
    AnimationFrame(data=(300, 300), duration=3),
    AnimationFrame(data=(500, 300), duration=6),
    AnimationFrame(data=(300, 500), duration=7)
])
circle_animator = Animator(circle, 'position', animation)


@window.event
def on_draw():
    window.clear()
    batch.draw()
    

pyglet.app.run()
```

**Comment by jason-bowen-zheng** [comment]
Thanks for example from @Kiera38, but we can also apply animates to several attributes like the example below:
```python
import pyglet

from pyglet.animation import Animator, Animation, AnimationFrame

window = pyglet.window.Window(960, 540)
batch = pyglet.graphics.Batch()
circle = pyglet.shapes.Circle(100, 100, 20, color=(200, 50, 50), batch=batch)

animation = Animation(frames=[
    AnimationFrame(r=30, position=(300, 300), duration=3),
    AnimationFrame(r=10, position=(500, 300), duration=6),
    AnimationFrame(r=20, position=(300, 500), duration=7)
])
circle_animator = Animator(circle, animation)


@window.event
def on_draw():
    window.clear()
    batch.draw()

@window.event
def on_mouse_press(*data):
    circle_animator.start()
    

pyglet.app.run()
```
We can also have methods to start and stop the animate.

**Comment by matanster** [comment]
I see that the google groups is very low traffic, so just linking here to a use case where animation support could avoid awkward ways of accomplishing the same. https://groups.google.com/g/pyglet-users/c/hxXXtkwcH0I/m/WidEiQKMBgAJ.  

I guess it's true as pointed out on the other issue thread pointing into here, that generic animation support is a slippery slope into animation pipelines and many other features that animation engines typically provide.

**Comment by matanster** [comment]
Well if it helps anyone more than being distracting, I wrote some code to bypass the stack nature of the current callbacks registration implementation. 

I hope this is concurrency safe as much as it seems to be in my own so far testing. Alas, maybe there would be re-entrancy issues if the overall application is dispatching functions through events and timers beyond what a very canonical pyglet application would do.

But otherwise, maybe it's a nice way to have animations and sound effects in a design scheme that isn't enabled in a push-pop design of callbacks.

In which case, no need for internal animations support in pyglet?

It is a class providing an interface for dynamically managing which callbacks would be invoked for each/any pyglet event type triggering callbacks, so that callbacks can be dynamically added and removed while multiple callbacks may be arbitrary registered and de-registered for the same pyglet event type.



```python
import contextlib
from typing import NamedTuple, Callable

import pyglet.window
import pytest

from nui.util.SimpleLogging import log

# an event type for which callbacks are being optionally invoked by pyglet,
# in pyglet source code parlance these are called event names, while we call
# them event type names in our code
EventTypeName = str

class CallbackRegistrationException(Exception):
    pass


class CallbackSpec(NamedTuple):
    method: Callable
    args: list = list()
    kwargs: dict = dict()

    def apply(self, *call_time_args):
        apply_args = self.args + list(call_time_args)
        return self.method(*apply_args)


class CallbackRegistry:

    def __init__(self, debug=True):

        self.debug = debug
        self._d: dict[EventTypeName, list[CallbackSpec]] = {}

        # while registering and deregistering callbacks, callback execution is deferred
        # (skipped or re-queued, depending on the event type), for which we need not a lock
        # and avoid a lock as it would cause deadlocks in our recursive calls where an
        # event callback may trigger its own deregistration """
        self.callbacks_execution_blocked = False

    @contextlib.contextmanager
    def callbacks_execution_integrity(self):
        """ a context manager that delays the execution of callbacks until the context is exited
        (as it is manipulating a binary variable, no thread locking is necessary as long as all
        uses of this variable are from python code (thanks to the GIL, see also
        https://stackoverflow.com/a/43007165/1509695 for motivation) """
        self.callbacks_execution_blocked = True
        yield
        self.callbacks_execution_blocked = False

    def register(self, et: EventTypeName, spec: CallbackSpec):
        """ registers a callback spec in the given callback event type name's list of callback specs """
        with self.callbacks_execution_integrity():
            if et in self._d:
                if spec in self._d[et]:
                    raise CallbackRegistrationException(f'{spec} already part of invocations list of {et}')
            else:
                self._d[et] = []
            self._d[et].append(spec)

    def deregister(self, et: EventTypeName, spec: CallbackSpec):
        """ removes a callback spec from the given callback event type name's list of registered callbacks """
        with self.callbacks_execution_integrity():
            if et in self._d:
                if spec in self._d[et]:
                    self._d[et].remove(spec)
                    if self._d[et] == []:
                        del self._d[et]
                else:
                    raise CallbackRegistrationException(f'{spec} not currently part of invocations list for {et}')
            else:
                raise CallbackRegistrationException(f'{spec} not currently part of invocations list for {et}')


    def register_many(self, callbacks_d: dict[EventTypeName, CallbackSpec]):
        """ registers one (or no) callback spec per callback event type name,
        intended for registering many callbacks when entering a state """
        with self.callbacks_execution_integrity():
            for et, spec in callbacks_d.items():
                if et in self._d:
                    if spec in self._d[et]:
                        raise CallbackRegistrationException(f'{spec} already part of invocations list of {et}')
                else:
                    self._d[et] = []
                self._d[et].append(spec)


    def deregister_many(self, callbacks_d: dict[EventTypeName, CallbackSpec]):
        """ removes one (or no) callback spec per callback event type name,
        intended for deregistering many callbacks when leaving a state """
        with self.callbacks_execution_integrity():
            for et, spec in callbacks_d.items():
                if et in self._d:
                    if spec in self._d[et]:
                        self._d[et].remove(spec)
                        if self._d[et] == []:
                            del self._d[et]
                else:
                    raise CallbackRegistrationException(f'{spec} not currently part of invocations list for {et}')


    def execute_callbacks(self, window: pyglet.window.Window, et: EventTypeName, *args):
        """ execute all callbacks for a given callback event type name, in their list order,
         unless changes to the callback registry are in progress, in which case for consistency
         (avoiding a functionally incoherent dictionary of callbacks being executed while being updated)
         we either skip execution of callbacks or delay them till after the registry update completes """
        if et in self._d:
            if self.callbacks_execution_blocked:
                # handle constantly firing callbacks differently than callbacks representing a one-time event
                if et in ['on_draw', 'on_update']:
                    log(f'{self.__class__.__name__} skipping execution of {et} callbacks while callbacks are being registered/deregistered', self.debug)
                else:
                    log(f'{self.__class__.__name__} delaying execution of {et} callbacks while callbacks are being registered/deregistered', self.debug)
                    window.dispatch_event(et, *args)
            else:
                for call_spec in self._d[et]:
                    call_spec.apply(*args)
```

Which hooks into the pyglet callback registration like so ...

```python
       @window.event
        def on_draw():

        @window.event
        def on_resize(new_width, new_height):
            self.registry.execute_callbacks(window, 'on_resize')

        @window.event
        def on_key_press(symbol, modifiers):
            self.registry.execute_callbacks(window, 'on_key_press', symbol, modifiers)

        @window.event
        def on_close():
            self.registry.execute_callbacks(window, 'on_close')

        def update(dt):
            self.registry.execute_callbacks(window, 'on_update')
```

Just a thought (and cursory implementation)

**Comment by caffeinepills** [comment]
In pyglet 3.0 development branch we have added further enhancements to the pyglet clock called "Event Chains" and this behavior should be possible now with those as tweens.
